Add documentation on testing Chrome Extensions (#7479)
GoogleChrome/developer.chrome.com: site/en/docs/extensions/mv3/end-to-end-testing/index.md

---
layout: "layouts/doc-post.njk"
title: "End-to-end testing Chrome Extensions"
seoTitle: "End-to-end testing Chrome Extensions"
date: 2023-10-09
description: How to write end-to-end tests for extensions.
---

End-to-end testing involves an extension package being built and loaded into a browser. A testing
tool communicates with the browser to automate interactions and test the same flows that a user
would go through. A library that supports end-to-end testing will generally provide ways of
controlling the browser, simulating user input and observing the current state of any open pages.

> See [Testing Chrome Extensions with Puppeteer][tutorial] for a tutorial.

Reviewer: ```suggestion
See [Testing Chrome Extensions with Puppeteer][tutorial] for a tutorial and [Unit testing][unit-testing] for details on writing unit tests for Chrome extensions.
```

## Using browser testing libraries {: #libraries }

Extensions are supported by a range of testing libraries.

| Library                | Guidance                                                                                                                              |

> |------------------------|:-------------------------------------------------------------------------------------------------------------------------------------:|

Reviewer: ```suggestion
|------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|
```

from the extension. In these cases, consider executing code in the context of an extension page.

In Puppeteer:

```js
const workerTarget = await browser.waitForTarget(
  target => target.type() === 'service_worker'
);
const worker = await workerTarget.worker();

const value = await worker.evaluate(() => {
  chrome.storage.local.get('foo');
});
```

In Selenium:

```js
// Selenium doesn't allow us to access the service worker, so we need to open an extension page where we can execute the code
await driver.get('chrome-extension://<id>/popup.html');
await driver.executeAsyncScript(
  'const callback = arguments[arguments.length - 1];' +
  'chrome.storage.local.get(\'foo\').then(callback);'
);
```

[puppeteer-testing]: https://pptr.dev/guides/chrome-extensions
[playwright-testing]: https://playwright.dev/docs/chrome-extensions
[selenium-chromeoptions]: https://www.selenium.dev/documentation/webdriver/browsers/chrome/
[selenium-extensions]: https://chromedriver.chromium.org/extensions
[webdriverio-testing]: https://webdriver.io/docs/extension-testing/web-extensions/
[new-headless]: /articles/new-headless/
[consistent-id]: /docs/extensions/mv3/manifest/key/#keep-consistent-id

> [tutorial]: /docs/extensions/mv3/tut_puppeteer-testing/

Reviewer: ```suggestion
[tutorial]: /docs/extensions/mv3/tut_puppeteer-testing/
[unit-testing]: /docs/extensions/mv3/unit-testing

```

---
layout: "layouts/doc-post.njk"

> title: "End-to-end testing Chrome Extensions"

Reviewer: ```suggestion
title: "End-to-end testing for Chrome Extensions"
```

seoTitle: "End-to-end testing Chrome Extensions"

> date: 2023-10-09

Reviewer: ```suggestion
date: 2023-10-12
```

description: How to write end-to-end tests for extensions.
---

End-to-end testing involves an extension package being built and loaded into a browser. A testing
tool communicates with the browser to automate interactions and test the same flows that a user
would go through. A library that supports end-to-end testing will generally provide ways of
controlling the browser, simulating user input and observing the current state of any open pages.

See [Testing Chrome Extensions with Puppeteer][tutorial] for a tutorial and [Unit testing][unit-testing] for details on writing unit tests for Chrome extensions.

## Using browser testing libraries {: #libraries }

Extensions are supported by a range of testing libraries.

| Library                | Guidance                                                                                                                              |

> |------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|

Reviewer: ```suggestion
|:------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|
```

---
layout: "layouts/doc-post.njk"
title: "End-to-end testing Chrome Extensions"
seoTitle: "End-to-end testing Chrome Extensions"
date: 2023-10-09
description: How to write end-to-end tests for extensions.
---

End-to-end testing involves an extension package being built and loaded into a browser. A testing
tool communicates with the browser to automate interactions and test the same flows that a user
would go through. A library that supports end-to-end testing will generally provide ways of
controlling the browser, simulating user input and observing the current state of any open pages.

See [Testing Chrome Extensions with Puppeteer][tutorial] for a tutorial.

> ## Using browser testing libraries

Reviewer: Does the guidance column look not quite right centered? 🤔

<img width="744" alt="Screenshot 2023-10-11 at 09 00 23" src="https://github.com/GoogleChrome/developer.chrome.com/assets/37001393/e409bbc2-5172-4edc-a0c5-7fb76d2b62d0">

Author: I agree, but I think this is just how tables work? I did try not using a table but I think this is much better than another level of headings.

Reviewer: To [left-align the column](https://www.markdownguide.org/extended-syntax/#alignment) you can replace the leftmost dash with a colon

Extensions are supported by a range of testing libraries.

| Library                | Guidance                                                                                                                              |
|------------------------|:-------------------------------------------------------------------------------------------------------------------------------------:|
| Puppeteer / Playwright | See Chrome Extensions ([Puppeteer][puppeteer-testing] / [Playwright][playwright-testing]).                                            |
| Selenium               | Use the [ChromeOptions][selenium-chromeoptions] object to load extensions. More information is available [here][selenium-extensions]. |
| WebDriverIO            | See [Web Extension Testing][webdriverio-testing].                                                                                     |

## Running tests in headless Chrome

> When running tests as part of an automated workflow, it is often necessary to load your extension on a machine that does not have a screen. Chrome's [new headless][new-headless] mode allows Chrome to be run in an unattended environment like this. Simply start Chrome with the `--headless=new` flag (headless currently defaults to "old", which does not support loading extensions). Depending on the automation tool you choose there may be a setting which adds the flag for you automatically.

Reviewer: ```suggestion
When running tests as part of an automated workflow, it is often necessary to load your extension on a machine that does not have a screen. Chrome's [new headless][new-headless] mode allows Chrome to be run in an unattended environment like this. Start Chrome using the `--headless=new` flag (headless currently defaults to "old", which does not support loading extensions). Depending on the automation tool you choose, there may be a setting that adds the flag for you automatically.
```

## Setting an extension ID

> It is often desirable to have a fixed extension ID in tests. This makes a number of common tasks easier such as allow-listing the extension's origin on a server it needs to communicate with, or opening extension pages within tests. To do this, follow the steps under [Keeping a consistent extension ID][consistent-id].

Reviewer: ```suggestion
It is often desirable to have a fixed extension ID in tests. This makes many common tasks easier, such as allow-listing the extension's origin on a server it needs to communicate with or opening extension pages within tests. To do this, follow the steps under [Keeping a consistent extension ID][consistent-id].
```

## Testing extension pages

Extension pages can be accessed using their corresponding URL, e.g `chrome-extension://<id>/index.html`. Use the normal navigation methods available in your automation tool of choice to navigate to these URLs.

## Testing an extension popup

Opening an extension popup in the context of another page is not currently possible. Instead, open the popup's URL in a new tab. If your popup uses the active tab, consider implementing an override where a tab ID can be passed explicitly to your popup. For example:

> 

Reviewer: Can we add a filename label here too?

Author: I added labels to all of the tutorial and places where the filename is important, but I don't think it makes sense here? This code would work anywhere and depends entirely on your project.

```js
const URL_PARAMS  = new URLSearchParams(window.location.search);

async function getActiveTab() {
  // Open popup.html?tab=5 to use tab ID 5, etc.
  if (URL_PARAMS.has("tab")) {
    return parseInt(URL_PARAMS.get("tab"));
  }

  const tabs = await chrome.tabs.query({
    active: true,
    currentWindow: true
  });

  return tabs[0];
}
```

## Inspecting extension state

To avoid test failures when you change the internal behavior of your extension, it is generally best
practice to avoid accessing internal state in an integration test. Instead, you should base your
tests on what is visible to the user. However, it can sometimes be desirable to directly access data
from the extension. In these cases, consider executing code in the context of an extension page.

In Puppeteer:

```js
const workerTarget = await browser.waitForTarget(
  target => target.type() === 'service_worker'
);
const worker = await workerTarget.worker();

const value = await worker.evaluate(() => {
  chrome.storage.local.get('foo');
});
```

In Selenium:

```js

> // Selenium doesn't allow us to access the background page, so we need to open an extension page we can execute the code in

Reviewer: ```suggestion
// Selenium doesn't allow us to access the service worker, so we need to open an extension page where we can execute the code in
```
